### PR TITLE
fixes to CurlHandler to pass on Alpine.39 Docker container

### LINKED ISSUE #36428: Test failure: System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandlerTest/GetAsync_IPv6LinkLocalAddressUri_Success on Alpine 3.9
Opened on behalf of @wfurt

The test `System.Net.Http.Functional.Tests.PlatformHandler_HttpClientHandlerTest/GetAsync_IPv6LinkLocalAddressUri_Success` has failed.

Failure Message:
```
System.Net.Http.HttpRequestException : An error occurred while sending the request.
---- System.Net.Http.CurlException : A libcurl function was given a bad argument
```

Stack Trace:
```
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts) in /__w/1/s/src/System.Net.Http/src/System/Net/Http/HttpClient.cs:line 500
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in /__w/1/s/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 107
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in /__w/1/s/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 107
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout) in /__w/1/s/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 71
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest.<>c__DisplayClass30_0.<<GetAsync_IPv6LinkLocalAddressUri_Success>b__0>d.MoveNext() in /__w/1/s/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:line 304
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options) in /__w/1/s/src/Common/tests/System/Net/Http/LoopbackServer.cs:line 69
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest.GetAsync_IPv6LinkLocalAddressUri_Success() in /__w/1/s/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:line 301
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace -----
   at System.Net.Http.CurlHandler.ThrowIfCURLEError(CURLcode error) in /__w/1/s/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.cs:line 660
   at System.Net.Http.CurlHandler.EasyRequest.SetUrl() in /__w/1/s/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs:line 282
   at System.Net.Http.CurlHandler.EasyRequest.InitializeCurl() in /__w/1/s/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs:line 86
   at System.Net.Http.CurlHandler.MultiAgent.ActivateNewRequest(EasyRequest easy) in /__w/1/s/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.MultiAgent.cs:line 723
```

Build : 3.0 - 20190327.6 (Core Tests)
Failing configurations:
- Alpine.39.Amd64-x64
  - Release

### PR DESCRIPTION
**Pull request by wfurt**
fixes #36428

There were actually two issues:
There is issue with Curl that it takes only _some_ values for  
```c# 
SetCurlOption(CURLoption.CURLOPT_ADDRESS_SCOPE, scopeId);
```
and it fails if scopeId is bigger than 0xF
https://github.com/curl/curl/blob/master/lib/setopt.c#L2302-L2312

That works on "normal" linux systems where eth0/eth1 would be bellow that as first interfaces. 
The troubles comes from docker where this may not be true:
```
bash-4.4$ ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
129: eth0@if130: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 02:42:ac:11:00:04 brd ff:ff:ff:ff:ff:ff
```
In my case eth0 has index and scopeId of 129.

It turned out this works when scopeId is encoded in URL. (as it probably should but we were afraid to do so as global change) 
Since that is corner case and low risk IMHO I added fixup code to do so. 

That fixes actual test failure described in #36428

The second part of this PR is dealing with the 
```
Assertion Failed
length 65524 should not be larger than RequestBufferSize 16384

```

It turned out that curl 7.62 added new API to set buffer size and it also increased the default to be bigger: https://github.com/curl/curl/commit/2825f46d950ca9b0e269ed8dc6a97c5238a17c30#diff-e69423595e1a7bd35e4b9868ca988550

Alpine 3.9 uses 7.64 and has this change. I don't think it is worth of setting up the size conditionally so I simple removed the buffer allowing to get buffer bigger than we need ( MaxRequestBufferSize)

With this all tests do pass on my local Alpine 3.9 instance using CI container image.  

This also explains failures in https://github.com/dotnet/corefx/pull/36392

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs [hunk 1]
```diff
@@ -263,22 +263,35 @@ private void SetUrl()
...
                     // Uri.AbsoluteUri doesn't include the ScopeId/ZoneID, so if it is link-local,
                     // we separately pass the scope to libcurl.
                     EventSourceTrace("ScopeId: {0}", scopeId);
-                    SetCurlOption(CURLoption.CURLOPT_ADDRESS_SCOPE, scopeId);
+                    try {
+                        SetCurlOption(CURLoption.CURLOPT_ADDRESS_SCOPE, scopeId);
+                    }
+                    catch (CurlException)
+                    {
+                        // If CURLOPT_ADDRESS_SCOPE becasue of curl bug, insert scopeId to url.
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs]: nit: spelling 'becasue' and grammar issue.

better would be:
```c#
// libCurl has a bug regarding long scope-ids:
//     https://github.com/curl/curl/issues/3713 
// Workaround the problem by percent-encoding the scope separator ('%')
```

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs [hunk 2]
```diff
@@ -263,22 +263,35 @@ private void SetUrl()
...
+                EventSourceTrace("Url: {0}", requestUri);
+                SetCurlOption(CURLoption.CURLOPT_PROTOCOLS, (long)(CurlProtocols.CURLPROTO_HTTP | CurlProtocols.CURLPROTO_HTTPS));
+
                 if (IsLinkLocal(requestUri, out scopeId))
                 {
                     // Uri.AbsoluteUri doesn't include the ScopeId/ZoneID, so if it is link-local,
                     // we separately pass the scope to libcurl.
                     EventSourceTrace("ScopeId: {0}", scopeId);
-                    SetCurlOption(CURLoption.CURLOPT_ADDRESS_SCOPE, scopeId);
+                    try {
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs]: nit: brace needs to be on next line

**Comment by wfurt** [comment]
This also fixes other Alpine and Fedora failures. They all now run inside of containers.
